src/common/schedulers/slurm_commands.py

```diff
@@ -112,36 +131,124 @@ def _recompute_required_nodes_per_job(pending_jobs, node_slots):
             # check the max number of slots I can fill with tasks of size cpus_min_per_node
             usable_slots_per_node = node_slots - (node_slots % job.cpus_min_per_node)
             # compute number of nodes by considering nodes of size usable_slots_per_node
-            required_nodes = int(math.ceil(job.cpus_total / usable_slots_per_node))
+            required_nodes = int(math.ceil(float(job.cpus_total) / usable_slots_per_node))
             # setting nodes to the max between the computed required nodes and the number of nodes given by
             # the squeue output. This is done to handle job submissions with the -N option where the user can
             # specify a number of nodes that is bigger than the min required.
             # e.g. sbatch -c 1 -N 2 -n 2 => (cpus_min_per_node=1, cpus_total=2, nodes=2)
             job.nodes = max(required_nodes, job.nodes)
 
 
+def _recompute_required_nodes_by_gpu_reservation(pending_jobs, gpus_per_node):
+    """
+    Adjust the number of required nodes if necessary based on the required GPUs.
+
+    When using --gpus option or --gpus-per-task without explicitly specifying the number of nodes Slurm will not
+    display the correct number of required nodes in the squeue output. The actual number needs to be recomputed
+    based on TRES_PER_JOB and TRES_PER_TASK data.
+
+    See examples below (Note output has been compressed):
+    ubuntu@ip-10-0-0-64:~$ sbatch --wrap "sleep 100" --gpus=12
+    Submitted batch job 79
+    ubuntu@ip-10-0-0-64:~$ /opt/slurm/bin/squeue -r -O 'jobid,statecompact,numnodes,numcpus,cpus-per-task,reason,
+        tres-per-job,tres-per-task'
+    JOBID|ST|NODES|CPUS|CPUS_PER_TASK|REASON|TRES_PER_JOB|TRES_PER_TASK
+    79|PD|1|1|1|ReqNodeNotAvail, Maygpu:12|N/A
+    ubuntu@ip-10-0-0-64:~$ sbatch --wrap "sleep 100" --gpus-per-task=2 -n 3
+    Submitted batch job 92
+    ubuntu@ip-10-0-0-64:~$ /opt/slurm/bin/squeue -r -O 'jobid,statecompact,numnodes,numcpus,cpus-per-task,reason,
+        tres-per-job,tres-per-task'
+    JOBID|ST|NODES|CPUS|CPUS_PER_TASK|REASON|TRES_PER_JOB|TRES_PER_TASK
+    92|PD|1|3|1|ReqNodeNotAvail, May|N/A|gpu:2
+
+    :param pending_jobs: array of SlurmJob
+    :param gpus_per_node: max gpus per compute node
+    """
+    if gpus_per_node <= 0:
+        # do not process any filtering since we assume scheduler rejects all jobs that require GPUs
+        return
+
+    for job in pending_jobs:
+        gpus_per_job = job.tres_per_job.get("gpu")
+        if gpus_per_job:
+            job.nodes = max(int(math.ceil(float(gpus_per_job) / gpus_per_node)), job.nodes)
+
+        gpus_per_task = job.tres_per_task.get("gpu")
+        if gpus_per_task:
+            tasks_schedulable_per_node = float(gpus_per_node) // gpus_per_task
+            job.nodes = max(int(math.ceil(float(job.tasks) / tasks_schedulable_per_node)), job.nodes)
```

Comment:
We also need to check/change `job.cpus_total` when we change `job.nodes` when looking at GPUs. For example: we have nodes with 4 GPUs on each node, we submit `sbatch -G 12`. In reality slurm will run this on minimum of 3 nodes, using 4 GPUs and 1 CPU on each node, so a total of 3 CPUs is required. Right now we will change `job.nodes` to 3 but we will not update `job.cpus_total` and that remains 1. This might cause a problem when we run get_required_nodes

Reply:
In other words I think we need to check `job.cpus_total >= job.node * cpus_min_per_node` after we update `job.nodes`

Reply:
Actually I can make this change in the next PR. Let's merge your modification first. Resolving this issue.